app/demo-geant-integration/HitRootIO.cc

```diff
@@ -0,0 +1,205 @@
+}  // namespace
+
+namespace demo_geant
+{
+G4ThreadLocal HitRootIO* HitRootIO::instance_ = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a ROOT output file for each worker except the master thread if MT
+ */
+HitRootIO::HitRootIO()
+{
+    G4AutoLock lock(&HitRootIOMutex);
+    file_name_ = std::regex_replace(
+        GlobalSetup::Instance()->GetSetupOptions()->output_file,
+        std::regex("json"),
+        "root");
+
+    if (G4Threading::IsWorkerThread())
+    {
+        file_name_ += std::to_string(G4Threading::G4GetThreadId());
+    }
+
+    if (G4Threading::IsWorkerThread()
+        || !G4Threading::IsMultithreadedApplication())
+    {
+        file_ = TFile::Open(file_name_.c_str(), "recreate");
+        CELER_VALIDATE(file_->IsOpen(), << "Failed to open " << file_name_);
+        tree_ = new TTree(
+            "Events", "Hit collections", this->SplitLevel(), file_);
+        G4AutoDelete::Register(tree_);
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the static thread local singleton instance
+ */
+HitRootIO* HitRootIO::GetInstance()
+{
+    if (instance_ == nullptr)
+    {
+        static G4ThreadLocalSingleton<HitRootIO> instance;
+        instance_ = instance.Instance();
+    }
+    return instance_;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write sensitive hits to output in the form of HitRootEvent.
+ */
+void HitRootIO::WriteHits(G4Event const* event)
+{
+    G4HCofThisEvent* HCE = event->GetHCofThisEvent();
+    if (HCE == nullptr)
+    {
+        return;
+    }
+
+    // Write the collection of sensitive hits into HitRootEvent
+    auto hit_event = std::make_unique<HitRootEvent>();
+    hit_event->SetEventID(event->GetEventID());
+    HitRootEvent::HitContainer* hcmap = hit_event->GetHCMap();
+    for (int i = 0; i < HCE->GetNumberOfCollections(); i++)
+    {
+        G4VHitsCollection* hc = HCE->GetHC(i);
+        G4String hcname = hc->GetName();
+        {
+            std::vector<G4VHit*> hits;
+            G4int number_of_hits = hc->GetSize();
+            for (G4int j = 0; j < number_of_hits; ++j)
+            {
+                G4VHit* hit = hc->GetHit(j);
+                SensitiveHit* sd_hit = dynamic_cast<SensitiveHit*>(hit);
+                hits.push_back(sd_hit);
+            }
+            hcmap->insert(std::make_pair(hcname, hits));
+        }
+    }
+
+    // Write a HitRootEvent into output ROOT file
+    G4AutoLock lock(&HitRootIOMutex);
+    this->WriteObject(hit_event.release());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Fill and write a HitRootEvent object
+ */
+void HitRootIO::WriteObject(HitRootEvent* hit_event)
+{
+    if (!init_branch_)
+    {
+        event_branch_ = tree_->Branch(
+            "event.",
+            &hit_event,
```

Comment:
This passes the address of temporary variable to the tree. The tree is retaining that address and will (with the current code) reuse it (incorrectly, leading to crashes) in subsequent call to this function.

Reply:
Changed as suggested.